File: gfg/POTD/2025-02-09/solution.java
/*
Node defined as
class Node{
    int data;
    Node left,right;
    Node(int d){
        data=d;
        left=right=null;
    }
}
*/
class Solution {
    // Function to return the maximum path sum from any node in a tree.
    public int calculateMaxPathSum(Node node, int[] globalMax) {
        if (node == null) return 0;
        
        int leftSubtreeSum = Math.max(calculateMaxPathSum(node.left, globalMax), 0);
        int rightSubtreeSum = Math.max(calculateMaxPathSum(node.right, globalMax), 0);
        
        globalMax[0] = Math.max(globalMax[0], leftSubtreeSum + rightSubtreeSum + node.data);
        
        return node.data + Math.max(leftSubtreeSum, rightSubtreeSum);
    }
    
    public int findMaxSum(Node root) {
        int[] globalMax = new int[1];
        globalMax[0] = Integer.MIN_VALUE;
        calculateMaxPathSum(root, globalMax);
        return globalMax[0];
    }
}

File: gfg/POTD/2025-02-09/solution.py

'''
class Node:
    def __init__(self,val):
        self.data = val
        self.left = None
        self.right = None
'''

class Solution:
    #Function to return maximum path sum from any node in a tree.
    def calculate_max_path_sum(self, node, global_max):
        if not node:
            return 0
        
        left_subtree_sum = max(self.calculate_max_path_sum(node.left, global_max), 0)
        right_subtree_sum = max(self.calculate_max_path_sum(node.right, global_max), 0)
        
        global_max[0] = max(global_max[0], left_subtree_sum + right_subtree_sum + node.data)
        
        return node.data + max(left_subtree_sum, right_subtree_sum)

    def findMaxSum(self, root):
        global_max = [float('-inf')]
        self.calculate_max_path_sum(root, global_max)
        return global_max[0]
